File: src/java/samples/foundgine-supply-chain-advanced/src/main/java/com/foundgine/samples/supplychain/advanced/authorization/SupplyChainAuthorization.java
package com.foundgine.samples.supplychain.advanced.authorization;

import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.authorization.*;
import com.foundgine.samples.supplychain.advanced.semantics.SupplyChainSemanticModel;
import java.util.*;

/** Supply-chain-specific policy data; authorization mechanics remain in Foundgine Core. */
public final class SupplyChainAuthorization {
    public enum Role { CUSTOMER, ANALYST, WAREHOUSE_OPERATOR, SUPPLY_CHAIN_MANAGER }

    public static ConfiguredSemanticAuthorizationPolicy create(String tenantId, Role role, Map<String,String> claims) {
        var context = new SemanticAuthorizationContext(tenantId, role.name(), claims);
        var config = new SemanticAuthorizationConfiguration()
                .addEntityRule((ctx, id, op) -> op == AuthorizationOperation.READ
                        ? canReadEntity(id, role) : canWriteEntity(id, role, ctx.safeClaims()))
                .addFieldRule((ctx, entity, field, op) -> op == AuthorizationOperation.READ
                        ? canReadField(entity, field, role) : canWriteField(entity, field, role, ctx.safeClaims()))
                .addRelationshipRule((ctx, entity, relationship, op) -> op == AuthorizationOperation.READ
                        ? canReadRelationship(relationship, role)
                        : !readOnly(ctx.safeClaims()) && role == Role.SUPPLY_CHAIN_MANAGER)
                .addPredicateRule(SupplyChainAuthorization::predicate)
                .addOperationRule((ctx, entity, op, name) -> namedOperation(ctx, op, name, role));
        return new ConfiguredSemanticAuthorizationPolicy(config, context);
    }

    public static FieldId field(String entity, String field) { return SupplyChainSemanticModel.field(entity, field); }
    public static RelationshipId relationship(String entity, String relationship) { return SupplyChainSemanticModel.relationship(entity, relationship); }

    private static boolean canReadEntity(EntityId id, Role role) {
        if (id.equals(SupplyChainSemanticModel.PRODUCT)) return true;
        if (id.equals(SupplyChainSemanticModel.SUPPLIER) || id.equals(SupplyChainSemanticModel.CERTIFICATION)
                || id.equals(SupplyChainSemanticModel.WAREHOUSE) || id.equals(SupplyChainSemanticModel.INVENTORY_LOT))
            return role != Role.CUSTOMER;
        if (id.equals(SupplyChainSemanticModel.COMPLIANCE_INCIDENT))
            return role == Role.ANALYST || role == Role.SUPPLY_CHAIN_MANAGER;
        return role != Role.CUSTOMER;
    }

    private static boolean canReadField(EntityId entity, FieldId field, Role role) {
        if (entity.equals(SupplyChainSemanticModel.INVENTORY_LOT)
                && field.equals(field("InventoryLot", "Quarantined")))
            return role == Role.WAREHOUSE_OPERATOR || role == Role.SUPPLY_CHAIN_MANAGER;
        if (entity.equals(SupplyChainSemanticModel.SUPPLIER)
                && field.equals(field("Supplier", "RiskScore")))
            return role == Role.ANALYST || role == Role.SUPPLY_CHAIN_MANAGER;
        return true;
    }

    private static boolean canReadRelationship(RelationshipId relationship, Role role) {
        if (relationship.equals(relationship("Supplier", "certifications"))) return role != Role.CUSTOMER;
        if (relationship.equals(relationship("Supplier", "incidents"))) return role == Role.ANALYST || role == Role.SUPPLY_CHAIN_MANAGER;
        if (relationship.equals(relationship("Warehouse", "inventory"))) return role != Role.CUSTOMER;
        return true;
    }

    private static boolean canWriteEntity(EntityId entity, Role role, Map<String,String> claims) {
        if (readOnly(claims)) return false;
        // Command entities are semantic capabilities backed by domain services.
        // The domain service remains the final enforcement point for actor/tenant
        // ownership; the semantic boundary must therefore allow the command itself.
        if (entity.equals(EntityId.create("PlaceOrderCommand"))
                || entity.equals(EntityId.create("CancelOrderCommand")))
            return true;
        return role == Role.WAREHOUSE_OPERATOR || role == Role.SUPPLY_CHAIN_MANAGER;
    }
    private static boolean canWriteField(EntityId entity, FieldId field, Role role, Map<String,String> claims) {
        if (readOnly(claims)) return false;
        if (entity.equals(EntityId.create("PlaceOrderCommand"))
                || entity.equals(EntityId.create("CancelOrderCommand")))
            return true;
        if (entity.equals(SupplyChainSemanticModel.INVENTORY_LOT))
            return field.equals(field("InventoryLot", "OnHand")) || field.equals(field("InventoryLot", "Reserved"));
        return role == Role.SUPPLY_CHAIN_MANAGER;
    }

    private static AuthorizationPredicate predicate(SemanticAuthorizationContext context, EntityId entity, AuthorizationOperation op) {
        AuthorizationPredicate result = null;
        if (op == AuthorizationOperation.READ && (entity.equals(SupplyChainSemanticModel.SUPPLIER)
                || entity.equals(SupplyChainSemanticModel.WAREHOUSE)
                || entity.equals(SupplyChainSemanticModel.INVENTORY_LOT)))
            result = tenantPredicate("TenantId");
        String warehouse = context.safeClaims().get("warehouse");
        if (op == AuthorizationOperation.READ && warehouse != null) {
            AuthorizationPredicate scope = null;
            if (entity.equals(SupplyChainSemanticModel.WAREHOUSE)) scope = warehousePredicate("Id", warehouse);
            else if (entity.equals(SupplyChainSemanticModel.INVENTORY_LOT)) scope = warehousePredicate("WarehouseId", warehouse);
            if (scope != null) result = result == null ? scope : AuthorizationPredicate.and(result, scope);
        }
        return result;
    }

    private static AuthorizationDecision namedOperation(SemanticAuthorizationContext context, AuthorizationOperation op,
                                                        AuthorizationOperationName name, Role role) {
        if (op != AuthorizationOperation.WRITE || name == null || !"inventory.reconcile".equalsIgnoreCase(name.value())) return null;
        if (role != Role.SUPPLY_CHAIN_MANAGER) return AuthorizationDecision.DENIED;
        return context.safeClaims().containsKey("reason") && context.safeClaims().containsKey("change_ticket")
                ? AuthorizationDecision.ALLOWED : AuthorizationDecision.DENIED;
    }

    private static boolean readOnly(Map<String,String> claims) { return "read-only".equalsIgnoreCase(claims.get("scope")); }
    private static AuthorizationPredicate tenantPredicate(String property) {
        return AuthorizationPredicate.equal(
                AuthorizationPredicate.member(AuthorizationPredicate.resourceParameter("resource"), property),
                AuthorizationPredicate.member(AuthorizationPredicate.contextParameter("context"), "TenantId"));
    }
    private static AuthorizationPredicate warehousePredicate(String property, String warehouse) {
        return AuthorizationPredicate.equal(
                AuthorizationPredicate.member(AuthorizationPredicate.resourceParameter("resource"), property),
                AuthorizationPredicate.constant(warehouse));
    }
    private SupplyChainAuthorization() {}
}

File: src/csharp/samples/Foundgine.SupplyChain.Advanced/Semantic/Authorization/SupplyChainAuthorization.cs
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Authorization;
using Foundgine.SupplyChain.Advanced.Semantics;

namespace Foundgine.SupplyChain.Advanced.Authorization;

public enum SupplyChainRole
{
    Customer,
    Analyst,
    WarehouseOperator,
    SupplyChainManager
}

/// <summary>
/// Sample-only application configuration. The authorization mechanics live in
/// Foundgine.Core.Semantic; this file contains only Supply Chain policy data and
/// actor-specific values.
/// </summary>
public static class SupplyChainAuthorization
{
    public static ConfiguredSemanticAuthorizationPolicy Create(
        string tenantId,
        SupplyChainRole role,
        IReadOnlyDictionary<string, string>? claims = null)
    {
        var context = new SemanticAuthorizationContext(tenantId, role.ToString(), claims);
        var config = new SemanticAuthorizationConfiguration()
            .AddEntityRule((ctx, id, operation) =>
                operation == AuthorizationOperation.Read
                    ? CanReadEntity(id, role)
                    : CanWriteEntity(id, role, ctx.SafeClaims))
            .AddFieldRule((ctx, entity, field, operation) =>
                operation == AuthorizationOperation.Read
                    ? CanReadField(entity, field, role)
                    : CanWriteField(entity, field, role, ctx.SafeClaims))
            .AddRelationshipRule((ctx, entity, relationship, operation) =>
                operation == AuthorizationOperation.Read
                    ? CanReadRelationship(entity, relationship, role)
                    : !IsReadOnly(ctx.SafeClaims) && role == SupplyChainRole.SupplyChainManager)
            .AddPredicateRule((ctx, entity, operation) => GetPredicate(ctx, entity, operation))
            .AddOperationRule((ctx, entity, operation, name) => GetNamedOperation(ctx, entity, operation, name, role));

        return new ConfiguredSemanticAuthorizationPolicy(config, context);
    }

    /// <summary>
    /// Authorization refers to semantic names and resolves generated identities
    /// through the discovered model. Numeric metadata IDs never become part of
    /// application policy source code.
    /// </summary>
    public static class FieldIds
    {
        public static FieldId SupplierRiskScore => SupplyChainSemanticModel.Field("Supplier", "RiskScore");
        public static FieldId InventoryOnHand => SupplyChainSemanticModel.Field("InventoryLot", "OnHand");
        public static FieldId InventoryReserved => SupplyChainSemanticModel.Field("InventoryLot", "Reserved");
        public static FieldId InventoryQuarantined => SupplyChainSemanticModel.Field("InventoryLot", "Quarantined");
    }

    public static class RelationshipIds
    {
        public static RelationshipId SupplierCertifications =>
            SupplyChainSemanticModel.Relationship("Supplier", "certifications");

        public static RelationshipId SupplierIncidents =>
            SupplyChainSemanticModel.Relationship("Supplier", "incidents");

        public static RelationshipId WarehouseInventory =>
            SupplyChainSemanticModel.Relationship("Warehouse", "inventory");
    }

    private static bool CanReadEntity(EntityId id, SupplyChainRole role) => id switch
    {
        var x when x == SupplyChainSemanticModel.Product => true,
        var x when x == SupplyChainSemanticModel.Supplier => role != SupplyChainRole.Customer,
        var x when x == SupplyChainSemanticModel.Certification => role != SupplyChainRole.Customer,
        var x when x == SupplyChainSemanticModel.ComplianceIncident => role is SupplyChainRole.Analyst
            or SupplyChainRole.SupplyChainManager,
        var x when x == SupplyChainSemanticModel.Warehouse => role != SupplyChainRole.Customer,
        var x when x == SupplyChainSemanticModel.InventoryLot => role != SupplyChainRole.Customer,
        _ => role != SupplyChainRole.Customer
    };

    private static bool CanReadField(EntityId entity, FieldId field, SupplyChainRole role)
    {
        if (entity == SupplyChainSemanticModel.InventoryLot && field == FieldIds.InventoryQuarantined)
            return role is SupplyChainRole.WarehouseOperator or SupplyChainRole.SupplyChainManager;
        if (entity == SupplyChainSemanticModel.Supplier && field == FieldIds.SupplierRiskScore)
            return role is SupplyChainRole.Analyst or SupplyChainRole.SupplyChainManager;
        return true;
    }

    private static bool CanReadRelationship(EntityId entity, RelationshipId relationship, SupplyChainRole role) =>
        relationship switch
        {
            var x when x == RelationshipIds.SupplierCertifications => role != SupplyChainRole.Customer,
            var x when x == RelationshipIds.SupplierIncidents => role is SupplyChainRole.Analyst
                or SupplyChainRole.SupplyChainManager,
            var x when x == RelationshipIds.WarehouseInventory => role != SupplyChainRole.Customer,
            _ => true
        };

    private static bool CanWriteEntity(EntityId _, SupplyChainRole role, IReadOnlyDictionary<string, string> claims) =>
        !IsReadOnly(claims) && role is SupplyChainRole.WarehouseOperator or SupplyChainRole.SupplyChainManager;

    private static bool CanWriteField(EntityId entity, FieldId field, SupplyChainRole role,
        IReadOnlyDictionary<string, string> claims)
    {
        if (IsReadOnly(claims)) return false;
        if (entity == SupplyChainSemanticModel.InventoryLot)
            return field == FieldIds.InventoryOnHand || field == FieldIds.InventoryReserved;
        return role == SupplyChainRole.SupplyChainManager;
    }

    private static AuthorizationPredicate? GetPredicate(SemanticAuthorizationContext context, EntityId entity,
        AuthorizationOperation operation)
    {
        AuthorizationPredicate? predicate = null;
        if (operation == AuthorizationOperation.Read &&
            (entity == SupplyChainSemanticModel.Supplier || entity == SupplyChainSemanticModel.Warehouse))
            predicate = TenantPredicate("TenantId");

        if (operation == AuthorizationOperation.Read && context.SafeClaims.TryGetValue("warehouse", out var warehouse))
        {
            var scope = entity == SupplyChainSemanticModel.Warehouse
                ? WarehouseIdPredicate("Id", warehouse)
                : entity == SupplyChainSemanticModel.InventoryLot
                    ? WarehouseIdPredicate("WarehouseId", warehouse)
                    : null;
            if (scope is not null)
                predicate = predicate is null ? scope : AuthorizationPredicate.And(predicate, scope);
        }

        return predicate;
    }

    private static AuthorizationDecision? GetNamedOperation(
        SemanticAuthorizationContext context,
        EntityId _,
        AuthorizationOperation operation,
        AuthorizationOperationName? name,
        SupplyChainRole role)
    {
        if (operation != AuthorizationOperation.Write || name is null ||
            !name.Value.Value.Equals("inventory.reconcile", StringComparison.OrdinalIgnoreCase))
            return null;

        if (role != SupplyChainRole.SupplyChainManager)
            return AuthorizationDecision.Denied;

        return context.SafeClaims.ContainsKey("reason") && context.SafeClaims.ContainsKey("change_ticket")
            ? AuthorizationDecision.Allowed
            : AuthorizationDecision.Denied;
    }

    private static bool IsReadOnly(IReadOnlyDictionary<string, string> claims) =>
        claims.TryGetValue("scope", out var scope) && scope.Equals("read-only", StringComparison.OrdinalIgnoreCase);

    private static AuthorizationPredicate TenantPredicate(string property) =>
        AuthorizationPredicate.Equal(
            AuthorizationPredicate.Member(AuthorizationPredicate.ResourceParameter("resource"), property),
            AuthorizationPredicate.Member(AuthorizationPredicate.ContextParameter("context"), "TenantId"));

    private static AuthorizationPredicate WarehouseIdPredicate(string property, string warehouseId) =>
        AuthorizationPredicate.Equal(
            AuthorizationPredicate.Member(AuthorizationPredicate.ResourceParameter("resource"), property),
            AuthorizationPredicate.Constant(warehouseId));
}